sql/catalyst/src/main/scala/org/apache/spark/sql/types/Decimal.scala

```diff
@@ -374,7 +374,7 @@ final class Decimal extends Ordered[Decimal] with Serializable {
       if (scale < _scale) {
         // Easier case: we just need to divide our scale down
         val diff = _scale - scale
-        val pow10diff = POW_10(diff)
+        val pow10diff = POW_10(math.min(diff, MAX_LONG_DIGITS))
```

Comment:
Can we still use compact long to store the decimal value? Shall we go to the branch `// Give up on using Longs; switch to BigDecimal, which we'll modify below`?

Reply:
Compact long can still be used when `diff > MAX_LONG_DIGITS`. In this case `longVal` becomes `0`, `1` or `-1` depending on rounding, so using `BigDecimal` is not actually needed.